client/service/src/builder.rs

```diff
@@ -650,6 +661,36 @@ impl<TBl, TRtApi, TCl, TFchr, TSc, TImpQu, TFprb, TFpp, TExPool, TRpc, Backend>
 			rpc_extensions,
 			remote_backend: self.remote_backend,
 			background_tasks: self.background_tasks,
+			block_announce_validator_builder: self.block_announce_validator_builder,
+			marker: self.marker,
+		})
+	}
+
+	/// Defines the `BlockAnnounceValidator` to use. `DefaultBlockAnnounceValidator` will be used by
+	/// default.
+	pub fn with_block_announce_validator(
+		self,
+		block_announce_validator_builder:
+			impl FnOnce(Arc<TCl>) -> Box<dyn BlockAnnounceValidator<TBl> + Send> + 'static,
```

Comment:
I think this should actually return a Result. That's how it's done in all the others and I guess it's for the same reasons than sc-cli's SubstrateCli having a lot of Result everywhere too

Reply:
The type is pretty complex here, maybe traitify this and then implement the trait for every `FnOnce() -> ... `?

Reply:
Also why isn't this just a generic type on `ServiceBuidler` level with static `new` function instead of a builder? Is there some dynamic state we expect to be transferred with the closure?
```rust
fn with_block_announce_validator<TBAV: BlockAnnounceValidator<TCl, TBl>>(self) -> ServiceBuilder<TBAV, TBl, ...> {
 ... // copy over the fields.
}

trait BlockAnnounceValidator<TCl, B> {
   fn new(client: Arc<TCl>) -> Self;
}
```

Reply:
No. There are 2 reasons for that:

1. In my use case for cumulus (see issue in description) we want to use a custom validator [JustifiedBlockAnnounceValidator](https://github.com/paritytech/cumulus/blob/b9793bc2f69ade2c7e2780d88c5dbedfc6adb7d0/network/src/lib.rs#L51) and as you can see the arguments for `new()` is actually something custom. I assume this can be personalized by the parachain as they want

2. On substrate [you can clearly see](https://github.com/paritytech/substrate/blob/a516cf292ef3293949d741a331c568446ce71b57/primitives/consensus/common/src/block_validation.rs#L56-LL66) that this has been done with this intention because the function `new` is not in the trait.

Reply:
To stay consistent with the existing API, I'd keep it this way for now.

Reply:
I thought that "staying consistent with the existing API" meant changing it to return a `Result`

Reply:
(Changing the closure to return a Result? Yes that's what I meant)